Packaging, installation, and testing

ceball opened the pull request:
* dependencies in one place (setup.py)
* `geoviews-core` (minimal dependencies) and `geoviews` packages (would correspond to `pip install geoviews` and `pip install geoviews[recommended]` if pip install were supported). `geoviews` (a metapackage) depends on `geoviews-core` (and is pinned to exact version).
* Packages are noarch:python
* package metadata read from setup.py
* travis will normally just test a 'dev environment'...
* ...but when a tag is pushed, will build, install, test, and - if tests successful - upload packages. Packages can be tested over multiple versions of python, with various combinations of dependencies, etc. Currently just run unit tests for geoviews-core and examples tests for geoviews, using python 2.7 and 3.6.
* conda package uploaded to anaconda.org pyviz when tag pushed (label/dev for pre-release tag)
* adds caching on travis for dev tests (the ones that run every time). Will have to see what impact this has on gv build times; can't tell yet as caching has not been on and/or has been messed up by changing stages round etc. Caching is disabled for packaging step - wouldn't have much effect anyway (conda build creates fresh test environments).
* adds testing on windows (appveyor)
* for testing notebooks, use nbsmoke (check for flakes, check they run without error)
* renamed `notebooks/` to `examples/`
* added geoviews command wrapping `geoviews.examples()` (`geoviews install_examples ...`) and `sample_data()`.
* fixed examples() by replacing with commands from pvutil (examples() was broken in last release when we created various subdirectories of notebooks).
* move unit tests into package, allowing installed package to be tested easily
* test commands all defined in tox.ini (and that's the only place they're defined). You don't like the cryptic structure of that file? Neither do I (https://github.com/pyviz/pyct/issues/9), but I think at least having the commands written in one place is a start.
* doit/pyctdev used to document tasks in a runnable way (runnable locally, on CI, across platforms). Work in progress. But all tasks just run standard tools, e.g. develop install is "conda install required and test dependencies, then `pip install -e . --no-deps`. Packaging uses conda build. Testing still uses nose. See see https://github.com/pyviz/pyctdev/blob/master/README.md for a draft overview of pyctdev.
* added pip package (though requires dependencies to be installed outside of pip). https://test.pypi.org/project/geoviews/
* simplified website build on travis. Also, new doc build uploaded to https://ioam-docs.github.io/geoviews-dev when a dev tag is pushed (if packages have successfully been built). New doc build uploaded to https://ioam-docs.github.io/geoviews-dev/ when a release tag is pushed (if packages have been successfully built). 


To do:
* [x] channels for package [edit: future work]
* [x] channels for dev build [edit: future work]
* [x] decide if geoviews-core and geoviews (recommended) packages is the way to go. It's like what e.g. dask does. But it's no work to revert to one package. Having the extra -core package allows someone to install without the extra dependencies (e.g. if geoviews were used as a library in other software).
* [x] [edit: restore `examples/` see hv for extending setuptools]; decide if `examples` symlink to `earthsim/examples` is ok. Negative: windows git developers may need to set up their git/system to support it. Negative: you may not like "seeing notebooks twice when you are operating in git repo". Positive: no code needed to copy files at package time. Positive: no code needed to find examples that depends on whether development or package install.
* [x] does `geoviews install_examples --include-data` fit in with future plans? Better just to release one more time with no command? (Not sure what future plans are. Maybe a pyviz command, which can install examples/download data for any pyviz project, available via the pyviz-cmd package?)
* [x] there are examples/assets that are actually examples/data ? [edit: future work]

To do for CB:
* [x] update pins for dependencies
* [x] not all notebooks run yet (probably combination of incomplete dependencies and data/assets in wrong location)
* [x] ...after that, go through list of dependencies and verify
* [x] restore coveralls EDIT: https://github.com/AnacondaRecipes/aggregate/issues/49
* [x] notebook flakes tests disabled; will make separate PR to remove notebook flakes
* [x] grep for TODOs and check they're done
* [x] update docs for installation EDIT: see [comment below](https://github.com/ioam/geoviews/pull/138#issuecomment-386416102)
* [x] pvutil: rename package, change command names, and then update here


-----

Original comment

Not ready to review (it's a mess), but @philippjfr I need some input from you. I think I want to know: what are the dependencies ideally/in theory (i.e. maybe just direct dependencies), and what are they in practice (i.e. maybe including indirect dependencies you need to pin)? And among the dependencies, what if anything is supposed to be optional (and for what 'feature'), and what's required?

I've made some combo/merge of the conda package dependencies and setup.py dependencies as currently on master:

```
+    python_requires = '>=2.7',
+    install_requires = [
+        'param >=1.5.1',
+        'numpy >=1.0',
+        'holoviews >=1.9.0',
+        'cartopy >=0.14.2',
+        'bokeh >=0.12.10',
+    ],
+    extras_require={
+        # Notebook dependencies of IPython
+        # TODO: is all this necessary? maybe just notebook, ipykernel or similar?
+        'notebook-dependencies' = [
+            'jupyter',
+            'pyzmq',
+            'jinja2',
+            'tornado',
+            'jsonschema',
+            'ipython',
+            'pygments',
+        ]},
```

I doubt that's going to be right. I think travis on master effectively does `conda install -c conda-forge nose numpy matplotlib bokeh pandas scipy jupyter ipython param flake8 mock filelock iris cartopy xarray geopandas numpy shapely=1.6.3 gdal=2.2.3 libgdal=2.2.3 glib=2.55.0 gstreamer=1.8.0 datashader` and `pip install git+https://github.com/ioam/holoviews.git`.

Also, do you know how the interaction between conda-forge and anaconda defaults works out for geoviews? Are there any problems with binary incompatibility for dependencies? I vaguely recall problems, but I don't remember clearly.

ceball commented:
Yes, the dependencies I tried are obviously not right. When conda build tries to import geoviews as a first test of the package, this happens:

<img width="1009" alt="screen shot 2018-02-17 at 11 47 09 am" src="https://user-images.githubusercontent.com/1929/36340584-60efe800-13d8-11e8-8738-c6999e48f69a.png">

I guess I need the just-released holoviews. I can adjust that one. But I'm sure there'll be other problems after, too...

philippjfr commented:
Current geoviews master requires current holoviews master. There has been no holoviews release containing trimesh yet.

ceball commented:
Ah, I got confused about the hv releases. Sorry!

It would still be useful to get an answer about the gv dependencies in general, though obviously no rush as it won't be possible to make a gv package without an hv one.

philippjfr commented:
I think jupyter or notebook is sufficient, other those dependencies look good. I'd maybe bump bokeh to >=0.12.13 though.

ceball commented:
* dependencies in one place (setup.py)
* `geoviews-core` (minimal dependencies) and `geoviews` packages (would correspond to `pip install geoviews` and `pip install geoviews[recommended]` if pip install were supported). `geoviews` depends on `geoviews-core` (and is pinned to exact version).
* Packages are noarch:python
* package metadata read from setup.py
* travis will normally just test a 'dev environment'...
* ...but when a tag is pushed, will build, install, test, and - if tests successful - upload packages. Packages can be tested over multiple versions of python, with various combinations of dependencies, etc. Currently just run unit tests for geoviews-core and examples tests for geoviews, using python 2.7 and 3.6.
* conda package uploaded to anaconda.org pyviz when tag pushed (label/dev for pre-release tag)
* adds caching on travis for dev tests (will have to see what impact this has on gv build times...)
* adds tests on windows
* for testing notebooks, use nbsmoke (check for flakes, check they run without error)
* added geoviews command (`geoviews install_examples ...`)
* move unit tests into package, allowing installed package to be tested easily
* renamed `notebooks/` to `examples/`
* test commands all defined in tox.ini (and that's the only place they're defined)
* fixed various flakes (until `doit test_flakes` stopped complaining).
* doit/pyct used to document tasks in a runnable way. Work in progress. But all tasks just run standard tools, e.g. develop install is "conda install required and test dependencies, then `pip install -e . --no-deps`. Packaging uses conda build. Testing still uses nose.


To do:
* [ ] update pins for dependencies
* [ ] not all notebooks run yet (probably combination of incomplete dependencies and data/assets in wrong location)
* [ ] ...after that, go through list of dependencies and verify
* [ ] channels for package: try just `-c ioam/label -c defaults`?
* [ ] channels for dev build: try just `-c ioam/label/dev -c defaults`? (except ioam/dev lags ioam/main?)
* [ ] geoviews/examples symlink is ok? (windows git developers may need to set up their git/system to support it; what about pip+git on windows? need to test) Otherwise, copy at package time?
* [ ] if symlink is ok, is symlink in repo is ok? (You may not like "seeing notebooks twice when you are operating in git repo".) Otherwise, create for duration of packaging?
* [ ] restore coveralls
* [ ] does `geoviews install_examples --include-data` fit in with future plans? Better just to release one more time with no command?
* [ ] flakes tests disabled; will make separate PR to remove flakes

philippjfr commented on setup.py:
```diff
@@ -58,37 +48,95 @@ def get_setup_version(reponame):
+        'xarray',
+        #### intent of packages below is jupyter notebook with ipykernel?
+        # seems like it could be simplified (or if for another
+        # purpose, should be clarified)
```
That's right, I don't think all those are needed, ``notebook`` is probably sufficient.

ceball replied:
It's probably notebook and ipykernel, or something like that. I've never tried to figure it out exactly...

philippjfr commented on setup.py:
```diff
@@ -62,9 +62,10 @@ def get_setup_version(reponame):
+# versions may be needed.
 extras_require={
     'recommended': [
+        'paramnb',
```
What's paramnb needed for?

philippjfr replied:
Just looked, ``notebooks/Dashboard_Demo.ipynb`` should just be deleted.

ceball replied:
Ok, have done that. 

What about Colocated_cube_pair.ipynb and CubeExplorer_Prototype.ipynb? Can they go onto the website, or they're not ready/suitable for website? (Either is fine, just checking.)

ceball replied:
At least on travis/linux, on this branch (where I've been changing the dependencies...), I get this:

Colocated_cube_pair:
<img width="1118" alt="screen shot 2018-04-30 at 11 54 16 am" src="https://user-images.githubusercontent.com/1929/39424217-6f6f21fc-4c6d-11e8-8ad2-ddf44eca3d70.png">

CubeExplorer_Prototype:
<img width="1080" alt="screen shot 2018-04-30 at 11 53 50 am" src="https://user-images.githubusercontent.com/1929/39424236-88970e60-4c6d-11e8-9897-3994c8c37529.png">

Haven't tried running them locally.

ceball commented:
I think updating the installation instructions doesn't need to be part of this PR, and should maybe be done later. Having said that, I would make them something like this...

-----

You can then install GeoViews and its other dependencies using conda:
   
```
conda install -c pyviz geoviews
```

You can now grab a copy of the notebooks to run locally, and run them using the Jupyter notebook:

```
geoviews examples --path geoviews-examples
cd geoviews-examples
jupyter notebook
```

If you want to try out the latest features between releases, you can get the latest dev release by specifying `-c pyviz/label/dev` in place of `-c pyviz`.

-----

Alternatively, if you are using conda-forge, you can also install iris and xesmf (TODO for what?); e.g.

```
conda create -n env-name -c pyviz -c conda-forge geoviews iris xesmf
conda activate env-name
```

**Note** Do not mix conda-forge and defaults: do not install packages from conda-forge into an environment created with defaults. If you are using the base environment of mini/anaconda, or an environment created without specifying conda-forge before defaults, and you then install from conda-forge, you will have incompatibilities in underlying, low-level dependencies and will experience segfaults.

philippjfr commented:
Looks like it's all finally working. Thanks so much for setting this up, really excited to start using it.
